DEVDOCS-4929: [update] Handlebars Helpers, add earlyHint helper (#1930)
bigcommerce/dev-docs: docs/stencil-docs/reference-docs/handlebars-helpers-reference.mdx

- [See it in GitHub](https://github.com/bigcommerce/paper-handlebars/blob/master/helpers/cdn.js)
- [See it in Cornerstone](https://github.com/bigcommerce/cornerstone/search?l=HTML&q=cdn)

#### Early Hint and cdn

> Early Hint reduces page load time and perceived latency by allowing browsers to download critical assets earlier in the request lifecycle. For more information, see [Early Hint](#earlyhint).

Reviewer: go ahead and make this plural to talk about them generally.

```suggestion
Early Hints reduce page load time and perceived latency by allowing browsers to download critical assets earlier in the request lifecycle. For more information, see [Early Hint](#earlyhint).

```

- [See it in GitHub](https://github.com/bigcommerce/paper-handlebars/blob/master/helpers/cdn.js)
- [See it in Cornerstone](https://github.com/bigcommerce/cornerstone/search?l=HTML&q=cdn)

#### Early Hint and cdn

> Early Hint reduces page load time and perceived latency by allowing browsers to download critical assets earlier in the request lifecycle. For more information, see [Early Hint](#early-hint).

Reviewer 2: ```suggestion
Early Hint reduces page load time and perceived latency by allowing browsers to download critical assets earlier in the request lifecycle. For more information, see [Early Hint](#earlyhint).

```

Reviewer 2: I believe the link to the heading is currently broken

You can view more Early Hints usage in Cornerstone's `templates/layout/base.html` file by visiting the [Cornerstone Repository](https://github.com/bigcommerce/cornerstone/commit/f70fcd502343503969c2e475b777a85347137542#).

### EarlyHint

```handlebars showLineNumbers

> {{earlyHint path preload}}

Reviewer 3: ```suggestion
{{ earlyHint '{path}' '{REL}' [as='{AS}'] [crossorigin='{CORS}']}}
```

You can view more Early Hints usage in Cornerstone's `templates/layout/base.html` file by visiting the [Cornerstone Repository](https://github.com/bigcommerce/cornerstone/commit/f70fcd502343503969c2e475b777a85347137542#).

### EarlyHint

```handlebars showLineNumbers

> {{ earlyHint '{path}' '{REL}' [as='{AS}'] [crossorigin='{CORS}']}}

Reviewer: ```suggestion
{{ earlyHint '{path}' '{rel}' [as='{AS}'] [crossorigin='{CORS}'] }}

```

```

Retrieves the provided URL and produces an early hint to communicate important web page resources. The server indicates which site assets (such as CSS files, JS scripts, and fonts) a browser needs to render a page fully. When used in conjunction with the `{{cdn}}` helper, load times can improve by 30%.

#### Parameters

- `path` (String): The provided path used when determining web page resources.

> - `REL` (String): Valid values are `preload`, `preconnect`, `prerender`,  and `dns-prefetch`.

Reviewer: ```suggestion
- `rel` (String): Valid values are `preload`, `preconnect`, `prerender`,  and `dns-prefetch`.

```

> - `AS` (String): Optional. Valid values are `script`, `style`, `font`.

Reviewer: ```suggestion
- `as` (String): Optional. Valid values are `script`, `style`, `font`.

```

> - `CORS` (String): Optional. Valid values are `no` (default), `anonymous`, and `use-credentials`.

Reviewer: ```suggestion
- `crossorigin` (String): Optional. Valid arguments are `no` (default), `anonymous`, and `use-credentials`.

```

Reviewer 3: This `Parameters` section confuses me because we provide an example of how the helper looks like with all its params (the snippet in lines 294-296), however, in this section we changed the placeholders: `AS`, and `CORS` by its attribute name. To make this more straightforward I'd change the aforementioned example by 

```
{{ earlyHint '{path}' '{rel}' [as='{script|style|font}'] [crossorigin='{no|anonymous|use-credentials}'] }}
```
Furthermore, I'd accept this suggestion above.

#### Examples

```handlebars showLineNumbers
{{earlyHint 'https://ajax.googleapis.com/ajax/libs/webfont/1.6.26/webfont.js' 'preload' as='script' crossorigin='anonymous'}}
```
This example produces an early hint in a `Link` header that looks like:

```handlebars showLineNumbers
<https://ajax.googleapis.com/ajax/libs/webfont/1.6.26/webfont.js>; rel=preload; as=script; crossorigin=anonymous
```

The next example includes additional helpers. `getFontsCollection` and `stylesheet` helpers produce early hints by default. 

> ```handlebars showLineNumbers

Reviewer: This example shows the cdn helper, but not the earlyHint helper.

Reviewer 3: That's correct. The case I tried to make here is to signal the helpers that produce early hints apart from `earlyHint`. I don't know if it would be better to signal those details in each helper section.

Reviewer: i think we just need to say that above, cc @bc-traciporter

You can view more Early Hints usage in Cornerstone's `templates/layout/base.html` file by visiting the [Cornerstone Repository](https://github.com/bigcommerce/cornerstone/commit/f70fcd502343503969c2e475b777a85347137542#).

### EarlyHint

```handlebars showLineNumbers

> {{ earlyHint '{path}' '{rel}' [as='{AS}'] [crossorigin='{CORS}'] }}

Reviewer: ```suggestion
{{ earlyHint '{path}' '{rel}' [as='{script|style|font}'] [crossorigin='{no|anonymous|use-credentials}'] }}
```

Reviewer: per @rafa-avila-bc

```

Retrieves the provided URL and produces an early hint to communicate important web page resources. The server indicates which site assets (such as CSS files, JS scripts, and fonts) a browser needs to render a page fully. When used in conjunction with the `{{cdn}}` helper, load times can improve by 30%.

#### Parameters

> - `path` (String): The provided path used when determining web page resources.

Reviewer: ```suggestion
- `path` (String): The URL that hosts the resource being hinted.

```
